State based fork implementation

tib888 opened the pull request:
The implementation of #285 is ready:
Similar functionality to the "mod-morph" of ZMK or to the "fork" of kanata, but more states can be tested: modifiers, mouse buttons, lock-leds* (documentation of keyboard config is included in PR).

Known problems:
- the led state based decisions do not work yet, since the received led states are not stored currently in the KeyboardReport.leds 

Please test, feedbacks are welcome!

Chaining many forks to test more complex conditions is coming soon...

tib888 commented:
The implementation of chainable forks is ready too.
Please review/accept the PR!

HaoboGu commented on rmk-macro/src/behavior.rs:
```diff
@@ -117,18 +118,190 @@ fn expand_combos(combos: &Option<CombosConfig>) -> proc_macro2::TokenStream {
+    //leds_compose: bool,
+    //leds_kana: bool,
+    //leds_power: bool,
+    //leds_shift: bool,
```
remove the commented lines if they're not used?

tib888 replied:
This part of StateBitsMacro goes to HidLeds, what is stored in keyboard.led_states: HidLeds, and sent to the PC in the KeyboardReport.

On the other hand, LightController receives HID reports with LedIndicator and based on that drives the led pins currently.

I think LightController **should also update** keyboard.led_states, but to do this, I'd like to ask your help.
(LedIndicator struct is the same as HidLeds, so only one of them should be kept). 

(If this would be ready, the forks would be able to change behavior based on these "lock leds", which would be very useful in my particular use-case...)

HaoboGu replied:
For led states, it worth a global value which could be accessed by both `Keyboard` and `LightController`. A `Mutex` might be needed

HaoboGu commented:
quite a complicated feature, please give me some time reviewing it .. 🥲

before that, could you run `cargo fmt` under the `rmk` folder to make all code refactored? thanks

tib888 commented:
It is cargo fmt already. mod.rs and rotary_encoder.rs needs formatting, but I haven't touched those. Should I format those too?

HaoboGu commented:
You don't have to. Great work! I'll take a deeper look asap

HaoboGu reviewed (commented):
Looks good overall! Could you please add some unittest also?

tib888 commented on rmk/src/storage/mod.rs:
```diff
@@ -228,6 +239,34 @@ impl Value<'_> for StorageData {
                 Ok(3 + COMBO_MAX_LENGTH * 2)
             }
+            StorageData::ForkData(fork) => {
+                if buffer.len() < 13 {
```
Fixed this.

tib888 commented:
> Looks good overall! Could you please add some unittest also?

I added some tests.

I run them with:
cargo test --tests --no-default-features --verbose
And some of them seem frozen (run more than 60s).

Also when I ran test_fork_with_held_mouse_button() test, it failed at ln 1831, while the same thing was fine on ln1768.
When I tried to debug, it was also fine. When I added a delay, it was also fine, but there is nothing time related in the test.
So, there is likely some strange async stuff happening...

tib888 commented:
I have traced down this freeze problem to:
KEYBOARD_REPORT_CHANNEL.sender().send(report).await

My guess is that the channel gets full if many test run. so this should be disabled during test runs somehow.

HaoboGu commented:
Yes you are correct. Running a dummy task which receives from `KEYBOARD_REPORT_CHANNEL` would solve this.

tib888 commented:
I let you do this, because I'm not an expert in these.
The nicest would be to be able to test in the unit tests, what hid reports sent into this dummy host...

HaoboGu commented:
Sure, would you please add all the test just as is? I'll add the dummy host stuffs.

tib888 commented:
> Sure, would you please add all the tests just as is? I'll add the dummy host stuffs.

Everything is already in the PR. Just uncomment the body of test_fork_with_held_modifier() (and test_fork_with_held_mouse_button())!

I do not want to push it uncommented because that leads to infinite test run on github.

HaoboGu commented:
thanks!
